src/wallet/wallet2.cpp

```diff
@@ -12519,6 +12519,37 @@ std::pair<uint64_t, std::vector<std::pair<crypto::key_image, crypto::signature>>
   return std::make_pair(offset, ski);
 }
 
+/**
+ *
+ * Erase corresponding incoming payments to the same subaddress account that is
+ * spending in a tx. "Payments" are supposed to be incoming payments from other
+ * users, or from the same user but different account. However, when scanning
+ * with just a view key and without either a spend key or knowing key images for
+ * received outputs (e.g. in a watch-only wallet that has not yet imported key
+ * images), when the scanner picks up a received output in a tx, it won't know
+ * if this received output is actually change or a self-spend, so it will look
+ * like an incoming "payment", and will incorrectly place it in the payments
+ * container. This function erases payments that are actually change, or are
+ * self-spends to the same account.
+ *
+ */
+void erase_incoming_payments_for_spent_tx(wallet2::payment_container &payments, const crypto::hash &spent_txid, uint32_t spending_subaddr_account)
+{
+  for (auto j = payments.begin(); j != payments.end(); )
+  {
+    if (j->second.m_tx_hash == spent_txid && j->second.m_subaddr_index.major == spending_subaddr_account)
```

Comment:
We're now requiring an extra condition to be filled for the payment to be erased: `j->second.m_subaddr_index.major == spending_subaddr_account`. So is it correct to say that now it won't be erased if sent to a different subaddress of the same account?

Reply:
Correct -- this is to match the behavior of the default scanner:

https://github.com/monero-project/monero/blob/9750e1fa103539b3e533455500610aae76e253a5/src/wallet/wallet2.cpp#L2399-L2404

Reply:
Sorry, slight correction: the payment won't be erased if sent to a **different account** than the spending account, from the same wallet

And this is to match behavior of the default scanner :)